File: src/main/java/dev/robothanzo/werewolf/listeners/MessageListener.java
package dev.robothanzo.werewolf.listeners;

import club.minnced.discord.webhook.WebhookClient;
import club.minnced.discord.webhook.send.WebhookMessage;
import club.minnced.discord.webhook.send.WebhookMessageBuilder;
import dev.robothanzo.werewolf.database.documents.Session;
import dev.robothanzo.werewolf.utils.CmdUtils;
import lombok.extern.slf4j.Slf4j;
import net.dv8tion.jda.api.entities.Webhook;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@Slf4j
public class MessageListener extends ListenerAdapter {
    public static Map<Long, WebhookClient> webhookCache = new HashMap<>();

    @NotNull
    public static WebhookClient getWebhookClientOrCreate(TextChannel channel) {
        if (!webhookCache.containsKey(channel.getIdLong())) {
            Webhook webhook = null;
            for (Webhook w : Objects.requireNonNull(channel.retrieveWebhooks().complete())) {
                webhook = w;
            }
            if (webhook == null) {
                webhook = Objects.requireNonNull(channel.createWebhook("Werewolf Impersonator").complete());
            }
            webhookCache.put(channel.getIdLong(), WebhookClient.withUrl(webhook.getUrl()));
        }
        return webhookCache.get(channel.getIdLong());
    }

    private boolean isCharacterAlive(Session session, String character) {
        for (Session.Player player : session.fetchAlivePlayers().values()) {
            if (player.getRoles() != null && player.getRoles().contains(character)) {
                // Check if this specific role is NOT dead
                if (player.getDeadRoles() == null || !player.getDeadRoles().contains(character)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean shouldSend(Session.Player player, Session session) {
        assert player.getRoles() != null;
        // Check first role (primary role for speech?) or any active role?
        // Logic seems to assume specific roles.
        // Assuming primary role is relevant, but with soft kill, roles order matters.
        // We should probably check if the role enabling speech is ALIVE.
        // But original code uses getRoles().getFirst().
        // If first role is dead in soft kill (but player alive), should he speak?
        // Probably not as that role.
        // But let's keep getFirst() for now, or check deadRoles.
        // Ideally we check if ANY of the enabling roles are alive.
        // But complicating this might break assumption that first role = current identity.
        // For now, let's just stick to replacement of getPlayers -> fetchAlivePlayers.
        return player.getRoles().getFirst().contains("狼人") ||
                player.getRoles().contains("狼兄") ||
                player.getRoles().getFirst().contains("狼王") ||
                player.getRoles().getFirst().contains("狼美人") ||
                player.getRoles().getFirst().contains("血月使者") ||
                player.getRoles().getFirst().contains("惡靈騎士") ||
                player.getRoles().getFirst().contains("夢魘") ||
                (player.getRoles().contains("狼弟") && !isCharacterAlive(session, "狼兄"));
    }

    @Override
    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
        if (event.getAuthor().isBot()) return;
        Session session = CmdUtils.getSession(event.getGuild());
        if (session == null) return;
        for (Session.Player player : session.fetchAlivePlayers().values()) {
            if (player.getRoles() != null && !player.getRoles().isEmpty()) {
                if (shouldSend(player, session) && player.getChannelId() == event.getChannel().getIdLong()
                        || event.getChannel().getIdLong() == session.getJudgeTextChannelId()) {
                    WebhookMessage message = new WebhookMessageBuilder()
                            .setContent(event.getMessage().getContentRaw())
                            .setUsername((((event.getChannel().getIdLong() == session.getJudgeTextChannelId()) ? "法官頻道" : player.getNickname())) +
                                    " (" + event.getAuthor().getName() + ")")
                            .setAvatarUrl(event.getAuthor().getAvatarUrl())
                            .build();
                    for (Session.Player p : session.fetchAlivePlayers().values()) {
                        if (shouldSend(p, session) && event.getChannel().getIdLong() != p.getChannelId()) {
                            getWebhookClientOrCreate(Objects.requireNonNull(event.getGuild().getTextChannelById(p.getChannelId()))).send(message);
                        }
                    }
                    if (event.getChannel().getIdLong() != session.getJudgeTextChannelId())
                        getWebhookClientOrCreate(Objects.requireNonNull(event.getGuild().getTextChannelById(session.getJudgeTextChannelId()))).send(message);
                    break;
                }
                if ((player.isJinBaoBao()
                        && player.getChannelId() == event.getChannel().getIdLong())) {
                    WebhookMessage message = new WebhookMessageBuilder()
                            .setContent(event.getMessage().getContentRaw())
                            .setUsername(player.getNickname() + " (" + event.getAuthor().getName() + ")")
                            .setAvatarUrl(event.getAuthor().getAvatarUrl())
                            .build();
                    for (Session.Player p : session.fetchAlivePlayers().values()) {
                        if (p.isJinBaoBao() && event.getChannel().getIdLong() != p.getChannelId()) {
                            getWebhookClientOrCreate(Objects.requireNonNull(event.getGuild().getTextChannelById(p.getChannelId()))).send(message);
                        }
                    }
                    break;
                }
            }
        }
    }
}

File: src/main/kotlin/dev/robothanzo/werewolf/listeners/MessageListener.kt
package dev.robothanzo.werewolf.listeners

import club.minnced.discord.webhook.WebhookClient
import club.minnced.discord.webhook.send.WebhookMessageBuilder
import dev.robothanzo.werewolf.database.documents.Player
import dev.robothanzo.werewolf.database.documents.Session
import dev.robothanzo.werewolf.service.GameSessionService
import dev.robothanzo.werewolf.service.NightManager
import dev.robothanzo.werewolf.utils.CmdUtils
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component

@Component
class MessageListener : ListenerAdapter() {
    @Autowired(required = false)
    @Lazy
    private lateinit var gameSessionService: GameSessionService

    @Autowired(required = false)
    @Lazy
    private lateinit var nightManager: NightManager

    companion object {
        val webhookCache: MutableMap<Long, WebhookClient> = HashMap()

        fun getWebhookClientOrCreate(channel: TextChannel): WebhookClient {
            if (!webhookCache.containsKey(channel.idLong)) {
                var webhook = channel.retrieveWebhooks().complete().firstOrNull()
                if (webhook == null) {
                    webhook = channel.createWebhook("Werewolf Impersonator").complete()
                }
                webhookCache[channel.idLong] = WebhookClient.withUrl(webhook.url)
            }
            return webhookCache[channel.idLong]!!
        }
    }

    private fun isCharacterAlive(session: Session, character: String): Boolean {
        for (player in session.alivePlayers().values) {
            if (player.roles?.contains(character) == true) {
                // Check if this specific role is NOT dead
                if (player.deadRoles == null || !player.deadRoles!!.contains(character)) {
                    return true
                }
            }
        }
        return false
    }

    private fun shouldSend(player: Player, session: Session): Boolean {
        // Kotlin handles nullability; assuming roles is not null per logic
        val roles = player.roles
        if (roles.isNullOrEmpty()) return false

        val firstRole = roles.filterNot { player.deadRoles?.contains(it) == true }.firstOrNull() ?: return false
        return firstRole.contains("狼人") ||
                roles.contains("狼兄") ||
                firstRole.contains("狼王") ||
                firstRole.contains("狼美人") ||
                firstRole.contains("白狼王") ||
                firstRole.contains("血月使者") ||
                firstRole.contains("惡靈騎士") ||
                firstRole.contains("夢魘") ||
                (roles.contains("狼弟") && !isCharacterAlive(session, "狼兄"))
    }

    override fun onMessageReceived(event: MessageReceivedEvent) {
        if (event.author.isBot) return
        val session = CmdUtils.getSession(event.guild) ?: return

        val alivePlayers = session.alivePlayers().values
        val isJudgeChannel = event.channel == session.judgeTextChannel

        for (player in alivePlayers) {
            if (!player.roles.isNullOrEmpty()) {
                if ((player.channel == event.channel && shouldSend(player, session)) ||
                    isJudgeChannel
                ) {
                    // Track werewolf messages during night phase for NightStatus
                    if (session.currentState == "NIGHT_PHASE" && ::gameSessionService.isInitialized) {
                        gameSessionService.withLockedSession(session.guildId) { lockedSession ->
                            // Record if it's from the judge channel or the werewolf's own channel
                            val messagesList = lockedSession.stateData.werewolfMessages

                            val senderId =
                                if (isJudgeChannel) 0 else lockedSession.getPlayer(event.author.idLong)?.id ?: 0
                            val displayName = event.member?.effectiveName ?: event.author.effectiveName
                            val senderName = if (isJudgeChannel) "法官頻道 (${displayName})" else displayName

                            // We should probably filter who sees these in the dashboard too, but for simplicity we'll keep all for now
                            // but mark them if possible.
                            messagesList.add(
                                dev.robothanzo.werewolf.game.model.WerewolfMessage(
                                    senderId = senderId,
                                    senderName = senderName,
                                    avatarUrl = event.author.effectiveAvatarUrl,
                                    content = event.message.contentRaw,
                                    timestamp = System.currentTimeMillis()
                                )
                            )

                            if (messagesList.size > 20) {
                                messagesList.removeAt(0)
                            }

                            // Notify NightManager and broadcast updated status inside lock to ensure data consistency
                            nightManager.notifyPhaseUpdate(lockedSession.guildId)
                            nightManager.broadcastNightStatus(lockedSession)
                        }
                    }

                    val message = WebhookMessageBuilder()
                        .setContent(event.message.contentRaw)
                        .setUsername(
                            (if (isJudgeChannel) "法官頻道" else player.nickname) +
                                    " (" + event.author.name + ")"
                        )
                        .setAvatarUrl(event.author.avatarUrl)
                        .build()

                    for (p in session.alivePlayers().values) {
                        if (shouldSend(p, session) && event.channel.idLong != p.channel?.idLong) {
                            val targetChannel = event.guild.getTextChannelById(p.channel?.idLong ?: 0)
                            if (targetChannel != null) {
                                getWebhookClientOrCreate(targetChannel).send(message)
                            }
                        }
                    }
                    if (event.channel != session.judgeTextChannel) {
                        session.judgeTextChannel?.let { getWebhookClientOrCreate(it) }?.send(message)
                    }
                    break
                }

                if (player.jinBaoBao && player.channel?.idLong == event.channel.idLong) {
                    val message = WebhookMessageBuilder()
                        .setContent(event.message.contentRaw)
                        .setUsername("${player.nickname} (${event.author.name})")
                        .setAvatarUrl(event.author.avatarUrl)
                        .build()

                    for (p in session.alivePlayers().values) {
                        if (p.jinBaoBao && event.channel.idLong != p.channel?.idLong) {
                            val targetChannel = event.guild.getTextChannelById(p.channel?.idLong ?: 0)
                            if (targetChannel != null) {
                                getWebhookClientOrCreate(targetChannel).send(message)
                            }
                        }
                    }
                    break
                }
            }
        }
    }
}
